File: generated_code/python/solution_3556_3.py
from typing import List
import heapq

class Solution:
    def getFinalState(self, nums: List[int], k: int, multiplier: int) -> List[int]:
        # Edge case: if multiplier is 1, operations don't change the values.
        if multiplier == 1:
            return [x % 1_000_000_007 for x in nums]
            
        MOD = 1_000_000_007
        n = len(nums)
        
        # Create a min-heap storing pairs of (value, original_index)
        # Python's heapq handles tuples by comparing the first element,
        # and if equal, comparing the second element (stable matching for earliest index).
        heap = [(val, i) for i, val in enumerate(nums)]
        heapq.heapify(heap)
        
        # Step 1: Simulate using the heap until the minimum element has been 
        # multiplied enough to become >= the original maximum element.
        # This aligns the elements so they maintain a predictable cycle.
        max_val = max(nums)
        
        while k > 0 and heap[0][0] * multiplier <= max_val:
            val, i = heapq.heappop(heap)
            new_val = val * multiplier
            heapq.heappush(heap, (new_val, i))
            k -= 1
            
        # If there are still remaining operations, we can distribute them efficiently.
        # At this point, the heap elements are ordered such that simulating 
        # exactly n operations would multiply every element by the multiplier exactly once.
        if k > 0:
            # Sort the heap elements to find their relative order after alignment.
            # Elements are sorted by current value, then by index.
            sorted_elements = sorted(heap)
            
            # Each element will complete at least (k // n) full cycles of multiplication.
            cycles = k // n
            rem = k % n
            
            # Compute multiplier^cycles % MOD using fast exponentiation
            pow_cycles = pow(multiplier, cycles, MOD)
            # Compute multiplier^(cycles + 1) % MOD for the remaining elements
            pow_cycles_plus_one = (pow_cycles * multiplier) % MOD
            
            # The first 'rem' elements in the sorted order get one extra multiplication.
            for i in range(n):
                val, idx = sorted_elements[i]
                if i < rem:
                    nums[idx] = (val % MOD) * pow_cycles_plus_one % MOD
                else:
                    nums[idx] = (val % MOD) * pow_cycles % MOD
        else:
            # If k reached 0 during the initial simulation, extract values from the heap.
            while heap:
                val, i = heapq.heappop(heap)
                nums[i] = val % MOD
                
        return nums

File: generated_code/cpp/solution_3556_3.cpp
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

class Solution {
private:
    // Helper function for fast modular exponentiation: (base^exp) % mod
    long long power(long long base, long long exp, long long mod) {
        long long res = 1;
        base %= mod;
        while (exp > 0) {
            if (exp % 2 == 1) res = (res * base) % mod;
            base = (base * base) % mod;
            exp /= 2;
        }
        return res;
    }

public:
    vector<int> getFinalState(vector<int>& nums, int k, int multiplier) {
        // Edge case: if multiplier is 1, array elements never change value
        if (multiplier == 1) {
            long long MOD = 1e9 + 7;
            vector<int> res(nums.size());
            for (int i = 0; i < nums.size(); ++i) {
                res[i] = nums[i] % MOD;
            }
            return res;
        }

        int n = nums.size();
        long long MOD = 1e9 + 7;

        // Min-heap stores pairs of {current_value, original_index}
        // Custom comparator ensures that if values are equal, the smaller index comes first
        auto comp = [](const pair<long long, int>& a, const pair<long long, int>& b) {
            if (a.first != b.first) return a.first > b.first;
            return a.second > b.second;
        };
        priority_queue<pair<long long, int>, vector<pair<long long, int>>, decltype(comp)> pq(comp);

        long long max_val = 0;
        for (int i = 0; i < n; ++i) {
            pq.push({nums[i], i});
            max_val = max(max_val, (long long)nums[i]);
        }

        // Phase 1: Simulation until the minimum element can reach or exceed the initial maximum.
        // This ensures all elements become relatively close in magnitude, establishing a cyclical order.
        while (k > 0 && pq.top().first * multiplier <= max_val * multiplier) {
            auto [val, idx] = pq.top();
            pq.pop();
            pq.push({val * multiplier, idx});
            k--;
        }

        // Phase 2: Mass processing using mathematical cycles.
        // Collect all elements from the heap into a sorted list based on their current states.
        vector<pair<long long, int>> sorted_elements;
        while (!pq.empty()) {
            sorted_elements.push_back(pq.top());
            pq.pop();
        }
        // Since pq is already a min-heap, popping all elements gives them in the exact order they will be picked.
        reverse(sorted_elements.begin(), sorted_elements.end()); 

        // Each element will get at least (k / n) complete cycles of multiplications.
        long long base_power = k / n;
        int remaining_ops = k % n;

        vector<int> final_state(n);
        for (int i = 0; i < n; ++i) {
            long long val = sorted_elements[i].first;
            int idx = sorted_elements[i].second;

            // The first 'remaining_ops' elements in the cycle get one extra multiplication.
            long long total_mults = base_power + (i < remaining_ops ? 1 : 0);

            long long mult_factor = power(multiplier, total_mults, MOD);
            final_state[idx] = ((val % MOD) * mult_factor) % MOD;
        }

        return final_state;
    }
};